### PR TITLE
Proxy support added to resolveWithPresence in HandledPromise

### PR DESCRIPTION
**Pull request by zarutian**
See [this comment](https://github.com/tc39/proposal-eventual-send/issues/8#issuecomment-579824005) for context.

### CODE REVIEW

**Comment by zarutian** [comment]
todo: add tests to excercise this.

**Comment by zarutian** [comment]
writing the fourth test is getting better of me, hence the numerious commits as I am doing this from a smartphone (local hw issues).

Well, @erights and @michaelfig take a look at this PR so far, please.

**Comment by michaelfig** [comment]
> writing the fourth test is getting better of me, hence the numerious commits as I am doing this from a smartphone (local hw issues).
> 
> Well, @erights and @michaelfig take a look at this PR so far, please.

Both @erights and I are preoccupied for now: him with vacation and me with https://hacktheorb.com. I haven't forgotten about you, though, and will get back to review sometime after next week.

**Comment by zarutian** [comment]
Well, this took a while and many small dot commits (those with a . at start of commit message)

I am ready for this PR to be reviewed.

Recommend that either @michaelfig or @erights do this review.

**Review by michaelfig — COMMENTED** [review]
I tried understanding the test but was wondering if you could explain the use-case for when you'd want to have a proxy?

This code looks reasonable to me.  I'll have to defer to @erights and @FUDCo as to whether it fits their model.

**Comment by zarutian** [comment]
Dynamic presences. Which are mainly usefull in implementing eventual send aware membranes. Example of such membrane is the whole captp subsystem that Agoric uses. Another such membrane is, yet to be written but planned, local-readonly-eventual-send-update membrane that encapsulates replicated statefull object graph. 

Not that much directly related to electronic rights but this package lives here.

**Comment by erights** [comment]
> This code looks reasonable to me. I'll have to defer to @erights and @FUDCo as to whether it fits their model.

Added myself as a reviewer.

**Review by erights — CHANGES_REQUESTED** [review]
Looks very close. Two issues:

I don't like having both a `revocable` flag and a `resultCallback`, where the `resultCallback` is really only needed in order to obtain the revoker. How about just a `revokerCallback` that, if present, causes the proxy to be a revocable one and is called back with the revoker? Let's discuss.

This package has an index.d.ts file describing its types. That file needs to be updated to reflect these new parameters.

File: packages/eventual-send/src/index.js
Comment (erights) [inline packages/eventual-send/src/index.js]:
```suggestion
        const { proxy: proxyOpts } = options;
        let presence;
        if (proxyOpts) {
          const {
            handler: proxyHandler,
            target: proxyTarget,
            revokerCallback,
          } = proxyOpts;
          if (revokerCallback) {
            // Create a proxy and its revoke function.
            const { proxy, revoke } = Proxy.revocable(proxyTarget, proxyHandler);
            presence = proxy;
            revokerCallback(revoke);
          } else {
            presence = new Proxy(proxyTarget, proxyHandler);
          }
        } else {
          // Default presence.
          presence = Object.create(null);
        }
```

Comment (erights) [reply packages/eventual-send/src/index.js]: As an experiment, I tried the above revision and then did a `yarn test` to find out what tests I'd need to adjust for this API change. All the tests still passed, which indicates that the previous API was not really being tested.

**Review by erights — DISMISSED** [review]

File: packages/eventual-send/src/index.js
Comment (zarutian) [reply packages/eventual-send/src/index.js]: This code was mostly adapted from Michaels code from the tc39/eventual-send issue linked in the first comment of this PR.

As someone that prefers implementing revocability oneself it had not occured to me to even test that when I was batttling with how to test presences with proxy. @michaelfig you recall why the resultCallback was needed for other than getting the revoker for revocable?

**Comment by zarutian** [comment]
> This package has an index.d.ts file describing its types. That file needs to be updated to reflect these new parameters.

I knew I forgot something.

File: packages/eventual-send/src/index.js
Comment (zarutian) [reply packages/eventual-send/src/index.js]: After looking at it and thinking a bit I do not see the any reason not to change resultCallback into revokerCallback as per thine suggestion.

**Comment by zarutian** [comment]
self-note: is: Það eina sem vantar í síðara revocable prófið er nokkur áköll eftir afturköllun og próf-yrðingar fyrir það.

**Comment by michaelfig** [comment]
@zarutian, are you blocked on our review?  I'd like to help get this merged if you think it is ready.

**Comment by zarutian** [comment]
I am terribly sorry. I had been intending to finish the test writing ages ago.
Would apriciate a little help there.

**Comment by michaelfig** [comment]
> I am terribly sorry. I had been intending to finish the test writing ages ago.

No worries! 🙂 

> Would apriciate a little help there.

I can help sometime when I'm doing work in this area.  Thanks for your contribution thus far!

**Comment by michaelfig** [comment]
Hi @zarutian, sorry for the wild goose chase, but I think this PR will be obsoleted by tc39/proposal-eventual-send#22.

Would you like to take a look at that API and see if it does what you need?  Your insight would be appreciated!

**Comment by zarutian** [comment]
What I am looking for is that presences can be proxies and can diffrientate between immediate calls and eventual sends, where it is the target.

I am looking at that tc39 issue at the moment.

**Comment by zarutian** [comment]
Looks like `presence` argument given to `resolveWithPresence` in [code quote 3 line 11](https://github.com/tc39/proposal-eventual-send/issues/22#issue-885531877) could be a Proxy. If so then my first criterion is met, looking if my second one is met or not.

**Comment by zarutian** [comment]
Both of my criterions  are met, per [answer](https://github.com/tc39/proposal-eventual-send/issues/22#issuecomment-850692414).

**Comment by michaelfig** [comment]
> Both of my criterions are met, per [answer](https://github.com/tc39/proposal-eventual-send/issues/22#issuecomment-850692414).

Cool.  However, it may be some time before we migrate to a new API, and I think your changes in the current PR are useful (especially due to the tests you added).  So, I'm going to merge the current PR (with my minor changes) as a step towards the new API.

**Review by michaelfig — APPROVED** [review]
LGTM!

**Pull request merged**
